### PR TITLE
Ast quantifier instantiations statistics

### PR DESCRIPTION
**Pull request by bobismijnnaam**
This PR enables quantifier statistics info printing from Z3. It also implements inline trigger extraction for forall and starall.

### CODE REVIEW

**Review by niomaster — APPROVED** [review]
Looks good! Some things to think about, but also happy to merge as is.

File: viper/src/main/java/viper/api/Silicon.scala [hunk 1]
```diff
@@ -31,8 +98,98 @@ case class Silicon(z3Settings: Map[String, String] = Map.empty, z3Path: Path = R
...
+    val reports = getQuantifierInstanceReports()
+    if (reports.nonEmpty) {
+      logger.info("=== Quantifier instantiation statistics ===")
+      val orderedReports = reports.sortBy(_.instances).reverse
+      val cutReports = if (logger.underlying.isDebugEnabled) orderedReports else orderedReports.take(10)
+      for (report <- cutReports) {
+        report.e match {
+          case Right(e) =>
+            val o = e.o
+            logger.info(s"${o.toString}: inst: ${report.instances} (gen: ${report.maxGeneration}, cost: ${report.maxCost})")
```
[the remark this answers is not in the document]
Comment (bobismijnnaam) [reply viper/src/main/java/viper/api/Silicon.scala:172]: Hmm, then it gets really minimal, as then you only get the start line and start column, right? I like the "filename.pvl: start - end" format, it's short but complete. I agree that using toString is not the right thing here. Maybe shortPosition can be changed, or maybe we need mediumPosition?

[the remark this answers is not in the document]
Comment (bobismijnnaam) [reply viper/src/main/java/viper/api/Silicon.scala:186]: Recapping the discussion: `shortPosition` was introduced for table error printing. We also determined that `shortPosition` is ambiguous anyway if multiple input files are used for VerCors. Hence, the table error printing mode is also ambiguous when multiple files are used. However, adding the full file path will make `shortPosition` not short. So we decided that only the base filename should change, as it is a nice tradeoff between unambiguousness and terseness.

File: viper/src/main/java/viper/api/Silicon.scala [hunk 2]
```diff
@@ -1,15 +1,48 @@
...
 @nowarn("any") // due to be removed
-case class Silicon(z3Settings: Map[String, String] = Map.empty, z3Path: Path = Resources.getZ3Path) extends SilverBackend {
+case class Silicon(z3Settings: Map[String, String] = Map.empty, z3Path: Path = Resources.getZ3Path, numberOfParallelVerifiers: Option[Int] = None,
+                   logLevel: Option[String] = None, proverLogFile: Option[Path] = None, printQuantifierStatistics: Boolean = false) extends SilverBackend {
+
+  var la: ConcurrentListAppender[ILoggingEvent] = null
+  var qmap: Map[String, Expr[_]] = Map()
+  var shutdownHookThread: ShutdownHookThread = null
+  var reportedQuantifiers = false
+  var intermediatePrinterTimer: Timer = new Timer()
```
[the remark this answers is not in the document]
Comment (bobismijnnaam) [reply viper/src/main/java/viper/api/Silicon.scala:49]: These two modules need to be more integrated indeed, for code reuse but also to make them similar in use. Let's do that in a next PR.

**Pull request merged**